### ISSUE TITLE
Finetuned motion control explanation request

### ISSUE THREAD

**Issue by tommyshelby4** [priorMDM/priorMDM#20]
As demonstrated in your repo, the results in the motion trajectory control are remarkable. For instance, the output figure writing a 'hello' in the air with unbelievable consistency. How is this specific sample output reproduced. Is the motion of the left wrist explicitly injected into either the 263-dimensional feature vector of the input or the xyz coordinates of the left-wrist of the input motion? In the same context, how is the root trajectory explicitly controlled and reproduced in the output? Lastly, what are the text prompts provided as supervision in each of the sample outputs you display in the [trajectory control section](https://priormdm.github.io/priorMDM-page/)?

**Comment by priorMDM** [comment]
Hello @tommyshelby4,
Thank your for the kind words!

The 263-dimensional feature includes the xyz coordinates of the left wrist (relative to the root location) in certain indices. So, in the mentioned example we injected the handcrafted left wrist xyz coordinates (relative to the root location) to those indices.
The root trajectory was also handcrafted and injected to its corresponding features in the 263-dimensional feature vector in a similar fashion.
Here we used the model blending method presented in the paper with both a model that was finetuned for wrist control and another model finetuned on trajectory control to achieve a motion that adheres to both control signals.
We hope in the future to release code that allow to conveniently use handcrafted motion features for controlling motion generation.

All the examples in the Fine-Tuned Motion Control section (including the trajectory control section) were generated with no text condition (note that the model is trained with occasional occlusion of the text condition as part of the classifier-free method so the model supports null-text generation out-of-the-box).

Hope this helps.

**Comment by tommyshelby4** [comment]
Thank you very much, both for the work and the time you devoted to respond!

**Comment by roykapon** [comment]
Thank you @tommyshelby4 for your interest! We used the linear velocities in x,z coordinates **and the root horizontal angle**. It is critical to also inject the root angle since the x,z velocities are relative to the root angle. For your case I guess you also wish to inject the y coordinate, but we only finetuned the model to adhere to the x,z coordinates. It could work fine without further finetuning but if it fails (generates motions that seem to not adhere to the y coordinate) you can solve it by finetuning the model to also adhere to the y coordinate of the root as well as the x,z coordinates (use --inpainting_mask root).

Also remember that when injecting your handcrafted features, you will probably inject them while the motion is being generated. Inside the denoising loop, the motion is represented in a normalized form. I.e., during training the model learns to generate motions that  are normalized by subtracting the mean value of each feature across the dataset and then dividing by the standard deviation. You can see this in  data_loaders/humanml/data
/dataset.py line 211, and can use the same code to normalize your handcrafted trajectory before injecting it.

Excuse me for the late response...

**Comment by shamangary** [comment]
This question has troubled me as well, here is an example of how to control the x,z speed (coordinate).

Please correct me if I am wrong @priorMDM 
```
    mean = np.load("./dataset/HumanML3D/Mean.npy")
    std = np.load("./dataset/HumanML3D/Std.npy")

    motion_len = 150
    model_kwargs["y"]["lengths"]  = model_kwargs["y"]["lengths"] * 0 + motion_len
    
    # Remember this is not absolute coordinate, this is speed/offset/vector.
    x_vec = (torch.ones(motion_len).unsqueeze(0).unsqueeze(0).to(input_motions) - mean[1])/std[1]
    z_vec = (torch.ones(motion_len).unsqueeze(0).unsqueeze(0).to(input_motions) - mean[2])/std[2]

    # input_motions: [batch, 263 dims, 1, seqlen]
    # erasing everything from input motion (optional)
    input_motions[:, :, :, :] = 0
    
    # x-axis
    input_motions[:, 1, :, :] = 0
    input_motions[:, 1, :, :motion_len] += x_vec
    # z-axis
    input_motions[:, 2, :, :] = 0
    input_motions[:, 2, :, :motion_len] += z_vec

    input_motions[:, :, :, motion_len:] = 0
    inpainting_mask[:, :, :, motion_len:] = 0
```

**Comment by wyyyz1937365497** [comment]
Is there anyone can teach me how to handcrafted and injected the 263-dimensional feature vector.
I have carefully read the code but still confused about how it work?
Thanks very much!
